services/kvstore/src/main/java/tech/pegasys/pantheon/services/kvstore/RocksDbKeyValueStorage.java

```diff
@@ -78,48 +80,53 @@ private RocksDbKeyValueStorage(
                   "read_latency_seconds",
                   "Latency for read from RocksDB.",
                   "database")
-              .labels(rocksDbConfiguration.getLabel());
+              .labels(dbLabel);
       removeLatency =
           metricsSystem
               .createLabelledTimer(
                   MetricCategory.KVSTORE_ROCKSDB,
                   "remove_latency_seconds",
                   "Latency of remove requests from RocksDB.",
                   "database")
-              .labels(rocksDbConfiguration.getLabel());
+              .labels(dbLabel);
       writeLatency =
           metricsSystem
               .createLabelledTimer(
                   MetricCategory.KVSTORE_ROCKSDB,
                   "write_latency_seconds",
                   "Latency for write to RocksDB.",
                   "database")
-              .labels(rocksDbConfiguration.getLabel());
+              .labels(dbLabel);
       commitLatency =
           metricsSystem
               .createLabelledTimer(
                   MetricCategory.KVSTORE_ROCKSDB,
                   "commit_latency_seconds",
                   "Latency for commits to RocksDB.",
                   "database")
-              .labels(rocksDbConfiguration.getLabel());
+              .labels(dbLabel);
 
       if (metricsSystem instanceof PrometheusMetricsSystem) {
-        RocksDBStats.registerRocksDBMetrics(stats, (PrometheusMetricsSystem) metricsSystem);
+        RocksDBStats.registerRocksDBMetrics(
+            stats, (PrometheusMetricsSystem) metricsSystem, dbLabel);
       }
 
-      metricsSystem.createLongGauge(
-          MetricCategory.KVSTORE_ROCKSDB,
-          "rocks_db_table_readers_memory_bytes",
-          "Estimated memory used for RocksDB index and filter blocks in bytes",
-          () -> {
-            try {
-              return db.getLongProperty("rocksdb.estimate-table-readers-mem");
-            } catch (final RocksDBException e) {
-              LOG.debug("Failed to get RocksDB metric", e);
-              return 0L;
-            }
-          });
+      metricsSystem
+          .createLabelledLongGauge(
+              MetricCategory.KVSTORE_ROCKSDB,
+              "rocks_db_table_readers_memory_bytes",
+              "Estimated memory used for RocksDB index and filter blocks in bytes",
+              "database")
+          .labels(dbLabel)
+          .accept(
```

Comment:
This isn't a particularly nice Api to expose. It's awkward having to chain the two calls to set labels and then create (with counters & timers you can defer supplying the labels until just before you update the metric but gauges have to supply both at construction).  And the `accept` method is probably overly generic here.

Three options I could see:
 1. `createLabelledGauge` returns something like `LabelledMetric<Gauge>` rather than using the generic `Consumer`.  `Gauge` would then have `createWithDataSource` (not exactly sure of the name maybe even just `dataSource` matching the style of `labels`) instead of `accept`.  Still a bit awkward but I think the meaningful method name makes it much more readable.
2. Don't try to shoehorn `LabelledMetric` in here - I don't think we get any real benefit from reusing it here. Then `createLabelledGauge` could return a `LabelledGauge` which could accept the label values and supplier in one call.
3. Do it all in one call, so `MetricSystem` exposes `createLabelledGauge` which takes label names, values and a supplier in one go.  Not quite sure how to make the API for that work well - it could take a Map of label name/values but it's a bit ugly on the calling side.

Not really a show stopper, just would be nice if we could make the creation of labelled gauges read a little nicer.

Reply:
I'm thinking of going down the `LabelledGauge` route, adding a 'remove gauge' option to it as well. The reason I went with the replace on create approach was that the pipeline tasks in the synchronizer are transient, so gauges come and go.  I couldn't run the second set of pipelines because the gauge already existed, unless I did a replace.  And unless we have a remove it also serves as a possible GC linkage from the metrics system to old and done eth tasks.  I'll iterate over it this week.